### PR TITLE
[WIP] Flatten branches when using one_of()

### PR DESCRIPTION
**Pull request by alexwlchan**
There’s probably a subtlety here that I’m missing, but let’s try this anyway.

When we get the arguments to `one_of()`, look through them to see if any of them are `OneOfStrategy`, then unpick the individual strategies from that. Then you have a flat list of strategies, as if you’d supplied them as individual arguments.

---

Here’s a program to show that this patch works as intended:

```python
from collections import Counter
from hypothesis.strategies import one_of, just

strategy = one_of(
    just(1),
    one_of(
        just(2),
        just(3),
        one_of(
            just(4),
            just(5),
            one_of(
                just(6),
                just(7)
            )
        )
    )
)

m = Counter([strategy.example() for _ in range(10000)])
print(m)
```

If you run that without the patch, we see that the inner strategies are heavily penalised in the distribution:

```
Counter({1: 4961, 3: 1676, 2: 1636, 5: 576, 4: 559, 7: 312, 6: 280})
```

Whereas if you run it with the patch, you see we have a much more even distribution:

```
Counter({3: 1496, 5: 1474, 1: 1440, 7: 1413, 4: 1404, 2: 1398, 6: 1375})
```

---

Assuming this approach isn’t completely silly, I’ll need some tests for this. I assume some tests similar to the sample program are the sort of thing we’d want – draw up a large sample, and test that the spread of the distribution isn’t too large. (And if I was more awake, I’d remember the appropriate bit of statistics, but I don’t.)

### CODE REVIEW

File: src/hypothesis/strategies.py [hunk 1]
```diff
@@ -173,16 +173,26 @@ def one_of(*args):
         except TypeError:
             pass
 
+    # If we have another one_of() in the list of arguments, pick out
+    # its underlying strategies
+    from hypothesis.searchstrategy.strategies import OneOfStrategy
+    strategies = []
```
Comment (alexwlchan) [inline src/hypothesis/strategies.py]: `strategies` is probably a terrible name there, because we haven’t called `check_strategy()` on these objects yet.

**Comment by DRMacIver** [comment]
I'm keen on this work! Thanks for starting on it.

I outlined an approach in https://github.com/HypothesisWorks/hypothesis-python/issues/234#issuecomment-237196449 that might work a little better than the isinstance check (in particular it commutes with map() and to some degree flatmap()), though if you want to stick to the isinstance check as a first pass I'm fine with that.

In terms of statistical tests, the Hypothesis test suite has your back here. You can just add things to [the statistical distribution tests](https://github.com/HypothesisWorks/hypothesis-python/blob/master/tests/nocover/test_statistical_distribution.py)

**Comment by alexwlchan** [comment]
So I’d add a `branches` property to the `SearchStrategy` class, which ordinarily returns a list containing itself, but in the case of `OneOfStrategy` or the map strategies, returns a list of multiple items? And use that to build up the list of intermediate strategies?

**Comment by DRMacIver** [comment]
Right, exactly.

**Comment by alexwlchan** [comment]
Okay, I’ve backed out the original approach and had a go using branch. This includes map() and flatmap(), I think, but not the deferred strategies.

New test case:

```python
double = lambda x: x * 2
strategy2 = one_of(
    just(1),
    one_of(
        (just(2) | just(3)).map(double),
        one_of(
            (just(4) | just(5)).map(double),
            one_of(
                (just(6) | just(7)).map(double)
            )
        )
    )
)
m = Counter(strategy2.example() for _ in range(10000))
print(dict(m))
```

Unpatched version:

```
$ python mytest.py
{1: 4927, 2: 1675, 3: 1702, 4: 555, 5: 577, 6: 300, 7: 264}
{1: 5053, 4: 1208, 6: 1283, 8: 589, 10: 644, 12: 605, 14: 618}
```

Patched version:

```
$ python mytest.py
{1: 1419, 2: 1382, 3: 1433, 4: 1480, 5: 1448, 6: 1406, 7: 1432}
{1: 1415, 2: 1482, 3: 1462, 4: 1439, 5: 1364, 6: 1392, 7: 1446}
```

Which is interesting, because I’ve got the distribution right, but dropped the `map()` somewhere along the way. Hmm.

File: src/hypothesis/strategies.py [hunk 2]
```diff
@@ -173,16 +173,18 @@ def one_of(*args):
         except TypeError:
             pass
 
+    strategies = []
     for arg in args:
         check_strategy(arg)
-    args = [a for a in args if not a.is_empty]
+        if not arg.is_empty:
+            strategies.extend(arg.branches)
```
Comment (alexwlchan) [inline src/hypothesis/strategies.py]: Hmm. I don't have commit access right now, but I wonder if I really want:

```python
strategies.extend([s for s in arg.branches it not s.is_empty])
```

**Comment by alexwlchan** [comment]
So the latest test failure is more interesting, and suggests I’ve introduced a subtle logic flaw somewhere in the `OneOfStrategy`:

```python
from hypothesis import strategies as st

x = st.integers()

# both return integers()
print(x | st.nothing())
print(st.one_of(x, st.nothing()))

# both return nothing()
print(st.nothing() | x)
print(st.one_of(st.nothing(), x))
```

Digging into this, it turns out this line is at fault; I should have `strategies` instead of `args`:

```python
if len(strategies) == 1:
    return args[0]
```

Correcting this error, I run into something slightly different: I now get `WideRangeIntStrategy()` for these three examples:

```python
print(st.one_of(x, st.nothing()))
print(st.nothing() | x)
print(st.one_of(st.nothing(), x))
```

(The first example has some logic in `__or__` that short-circuits passing to `one_of()`.)

Diving into the definition of `integers()`, I can see that `WideRangeIntStrategy()` is one of the possible return values from `integers()`, depending on the user-supplied bounds. It seems like something I’m doing here is “activating” the strategy, but I’m not entirely sure what it is:

```python
strategies = []
for arg in args:
    check_strategy(arg)
    if not arg.is_empty:
        strategies.extend([s for s in arg.branches if not s.is_empty])
```

Any ideas @DRMacIver?

**Comment by alexwlchan** [comment]
Ah, I have an idea: I added the following method to `DeferredStrategy`:

```python
@property
def branches(self):
    return self.wrapped_strategy.branches
```

Taking this out seems to resolve that issue *and* preserve the statistical distribution. Because `DeferredStrategy`, as the name implies, isn’t actually reifying the strategy into anything like `WideRangeIntStrategy()`, it’s just holding the `integers()` instance ready to be reified.

Let’s take that out and see what happens.

**Comment by DRMacIver** [comment]
Hmm. No, that should have been the correct implementation for DeferredStrategy. I think what you're seeing is actually mostly correct behaviour. The strategy that previously displayed as 'integers()' was probably just a thin shell over WideRangeIntegerStrategy that fixed up its repr.

It *would* be nice to keep reprs consistent and not expose internal names though. A cheap trick that might work but may have edge cases I'm currently missing is to have DeferredStrategy return (self,) if the child's branches has length one?

**Comment by alexwlchan** [comment]
I’ve pushed another version with that fix. It seems to get the repr() right in that case, at least. Let’s see what CI thinks.

**Comment by alexwlchan** [comment]
We’re down to the interesting failures now: coverage is complaining because I haven’t written any tests yet, and a pretty repr that’s broken, but only on Linux.

**Comment by alexwlchan** [comment]
Maybe not – I’ve just tried to run the same test on macOS, and it hits a failure:

```
_____________________________ test_repr_evals_to_thing_with_same_repr ______________________________
tests/nocover/test_pretty_repr.py:114: in test_repr_evals_to_thing_with_same_repr
    @settings(max_examples=2000)
src/hypothesis/core.py:524: in wrapped_test
    print_example=True, is_final=True
src/hypothesis/executors.py:58: in default_new_style_executor
    return function(data)
src/hypothesis/core.py:111: in run
    return test(*args, **kwargs)
tests/nocover/test_pretty_repr.py:119: in test_repr_evals_to_thing_with_same_repr
    assert r == r2
E   assert 'floats().map....denominator)' == 'floats().map(...f: <unknown>)'
E     - floats().map(float_to_decimal) | fractions().map(lambda f: Decimal(f.numerator) / f.denominator)
E     ?                                                            ^^^^^^^^^^ ^^^^^^^^^^^^^^^^  ^^ ^^^^
E     + floats().map(float_to_decimal) | fractions().map(lambda f: <unknown>)
E     ?                                                            ^^ ^  ^ ^
```

**Comment by DRMacIver** [comment]
Huh. The lambda printing is a weird thing for this change to have affected (though I haven't actually looked at the diff).

I'm a bit obsessed with pretty reprs, but given that there's a known fix for all incorrect displays of <unknown> that is currently reverted due to licensing reasons you should feel free to just change the test if it proves a massive pain to fix that one.

**Comment by alexwlchan** [comment]
Lambda printing is bizarre.

If I extract that test case and the example into a standalone file, and run it, I get the same bad repr on my branch or with a release version of Hypothesis. In a clean virtualenv with Hypothesis installed, this is the output:

```console
$ python mytest3.py
floats().map(float_to_decimal) | fractions().map(lambda f: Decimal(f.numerator) / f.denominator)
floats().map(float_to_decimal) | fractions().map(lambda f: <unknown>)
Traceback (most recent call last):
  File "mytest3.py", line 19, in <module>
    assert r == r2
AssertionError
```

Okay, so maybe I should run the tests against master with the version of the Hypothesis database that knows about this error? All pass fine.

Maybe something’s gone wrong; I’ll provide it as an explicit example:

```python
@given(Strategies)
@settings(max_examples=2000)
@example(strategy=st.floats().map(st.float_to_decimal) | st.fractions().map(lambda f: Decimal(f.numerator) / f.denominator))
def test_repr_evals_to_thing_with_same_repr(strategy):
    r = repr(strategy)
    via_eval = eval(r, strategy_globals)
    r2 = repr(via_eval)
    assert r == r2
```

Now the test blows up with a SyntaxError, because the first `repr()` has added an `<unknown>`, which chokes the `eval()`. In the standalone case, the first repr was correct, and it was the repr of the eval'd version that added `<unknown>`.

I wonder why they’re different in a standalone/test environment.

---


Although weirdly, the problem doesn’t get flagged in the test runner.

The problem with the lambda appears to come from inside `reflection.py`:

```python
try:
    source = inspect.getsource(f)
except IOError as exc:
    print('@@AWLC IOError', exc)
    return if_confused
```

where we get the error "could not get source code". If this is a bug in Hypothesis master, I wonder why we haven’t spotted it before.

Also weird: the repr() is different

<details>
<summary>Test case</summary>
```python
#!/usr/bin/env python
# -*- encoding: utf-8 -*-

from decimal import Decimal
import hypothesis.strategies as st

strategy_globals = dict(
    (k, getattr(st, k))
    for k in dir(st)
)

strategy=st.floats().map(st.float_to_decimal) | st.fractions().map(lambda f: Decimal(f.numerator) / f.denominator)

r = repr(strategy)
print(r)
via_eval = eval(r, strategy_globals)
r2 = repr(via_eval)
print(r2)
assert r == r2
```
</details>

**Comment by DRMacIver** [comment]
I think whenever you're seeing the repr be different that's because the change you've made has resulted in a strategy that was previously being wrapped becoming unwrapped.

A thing that might make sense is to pass the unflattened strategies to OneOfStrategy, have it do the flattening, but use the original strategies for the repr?

It would be nice to get to the bottom of these issues though.

**Comment by alexwlchan** [comment]
Hooray! The tests are basically passing, modulo lint and coverage problems. So I think this leaves:

* [ ] Fix lint errors
* [ ] Add tests that get up to 100% coverage
* [ ] Add some statistical distribution tests
* [ ] Rebase into something vaguely coherent

@DRMacIver I probably won’t finish this before you go on your road trip, so I’ll probably finish it off and leave it sitting in a branch for when you get back.

**Comment by DRMacIver** [comment]
I'm leaving in the next six hours, so that's very likely! 

I may however review it while on the road (well not literally). I don't intend to be completely incommunicado.

**Comment by DRMacIver** [comment]
What's the status of this?

If you don't have time/inclination to work on it right now that's obviously 100% fine. I'm just checking because I'm keen on this work and wanted to make sure you weren't blocked on something I could help with.

**Comment by alexwlchan** [comment]
Sorry, real life got in the way a bit. I'm planning to look at it again in the next few days, will shout if I run into issues.

**Comment by DRMacIver** [comment]
> Sorry, real life got in the way a bit

This is totally not a thing you need to apologise for!

**Comment by alexwlchan** [comment]
So I seem to be getting a bunch of warnings from pytest, which is weird because I don’t think I’ve been near these classes. Any suggestions for what I’m doing wrong?

```
WC1 /home/travis/build/HypothesisWorks/hypothesis-python/tests/cover/test_conjecture_engine.py cannot collect test class 'TestData' because it has a __init__ constructor
WC1 /home/travis/build/HypothesisWorks/hypothesis-python/tests/cover/test_conjecture_engine.py cannot collect test class 'TestRunner' because it has a __init__ constructor
WC1 /home/travis/build/HypothesisWorks/hypothesis-python/tests/cover/test_conjecture_test_data.py cannot collect test class 'TestData' because it has a __init__ constructor
WC1 /home/travis/build/HypothesisWorks/hypothesis-python/tests/cover/test_conjecture_utils.py cannot collect test class 'TestData' because it has a __init__ constructor
WC1 /home/travis/build/HypothesisWorks/hypothesis-python/tests/cover/test_integer_ranges.py cannot collect test class 'TestData' because it has a __init__ constructor
```

**Comment by DRMacIver** [comment]
Those warnings aren't your fault and are a pre-existing thing that I haven't got around to to doing something about because they're harmless and I'm being lazy about it.

The easy solution is probably to change those names (which are internal API only)

**Comment by alexwlchan** [comment]
Replaced by #406.

**Pull request closed without merging**